### ISSUE TITLE
[ERTP] Difficulties working in multi-vat scenarios

### ISSUE THREAD

**Issue by DavidBruant** [Agoric/agoric-sdk#708]
While working on #615, i'm coming across cases where ERTP functions do not work in cross-vat contexts

For instance, i'm getting this error:
```sh
FAILED ASSERTION false
payment not found: Promise <Promise <[Object: null prototype] {}>> [Promise] { <pending> }
    doProcess: vat[v4][o+4].getAmountOf dispatch failed: Error <Object <[Object: null prototype] {}>>: payment not found: (an object)
```

When i track down the error, i see that it comes from the connection of the following places:
https://github.com/Agoric/agoric-sdk/blob/9d8b545f6f8d0372bc83a33db7dc9d6ee87a8027/packages/weak-store/src/weakStore.js#L20-L30

https://github.com/Agoric/agoric-sdk/blob/9d8b545f6f8d0372bc83a33db7dc9d6ee87a8027/packages/ERTP/src/issuer.js#L21

https://github.com/Agoric/agoric-sdk/blob/9d8b545f6f8d0372bc83a33db7dc9d6ee87a8027/packages/ERTP/src/issuer.js#L158

My best guess would be that the promise is not well-recognized or not well-unwrapped before it's passed to `getAmountOf`/`paymentLedger.get`

The unit tests do not test cross-vat contexts: https://github.com/Agoric/agoric-sdk/blob/master/packages/ERTP/test/unitTests/test-issuerObj.js

There is a test in [multi-vat context](https://github.com/Agoric/agoric-sdk/blob/master/packages/ERTP/test/swingsetTests/splitPayments/bootstrap.js), but the issuer is part of the boostrap "vat" ("context" ?). I wonder whether this makes a difference with spawner tests where even the mint and issuers are born in a dedicated vat and the only used cross-vat:

https://github.com/Agoric/agoric-sdk/blob/9d8b545f6f8d0372bc83a33db7dc9d6ee87a8027/packages/spawner/test/swingsetTests/contractHost/bootstrap.js#L181-L212

---

Among the solutions that can be created at various places and potentially conflicts with one another:
- it'd be cool if ERTP caught the weak-store error and provided a higher-level error
- it's not clear to me whether weak-store and/or ERTP and/or the user should be doing the unwrapping
I see that some ERTP methods do it, but it's inconsistent:

https://github.com/Agoric/agoric-sdk/blob/9d8b545f6f8d0372bc83a33db7dc9d6ee87a8027/packages/ERTP/src/issuer.js#L44

https://github.com/Agoric/agoric-sdk/blob/9d8b545f6f8d0372bc83a33db7dc9d6ee87a8027/packages/ERTP/src/issuer.js#L157-L160

- it might make sense to add a comprehensive multi-vat test suite in ERTP, because it likely will be used a lot like this
- is it time to remove `E` and write `~.` code everywhere right now?

**Comment by katelynsills** [comment]
Ah! You are right that `getAmountOf` was missed. I can create a PR for that on master, unless you want to take that on.

I believe that every single ERTP method that can take a payment should be able to take a promise for a payment. Can you help me make sure that that is the case? 

> it might make sense to add a comprehensive multi-vat test suite in ERTP, because it likely will be used a lot like this

Yes - we previously had some of that in the form of Zoe swingset tests that relied heavily on ERTP, but when Zoe was moved out into its own package, it became clear that we did not in fact have a good number of Swingset tests for ERTP itself. I think what might have happened in this case is that with `makePromise` not handling things correctly from the beginning, I probably `await`-ed from the user-side, thus not catching this. 

> is it time to remove E and write ~. code everywhere right now?

Not until we have VS Code linting support.

**Comment by katelynsills** [comment]
Responding to https://github.com/Agoric/agoric-sdk/pull/621#discussion_r393857814

`getAmountOf` should accept a promise for a payment. The fact that it doesn't was an oversight, so this was a great thing to discover. Thank you, @DavidBruant!

> is a presence a promise?

A presence is not a promise. A presence is a local object that is representing a remote object - any method calls on the local object will actually be sent to the remote object. However, any inter-vat calls will always result in a promise (which may be for a presence). 

> can a presence represent a promise in another vat? (or should it always represent a non-promise object)

As far as I know, it is only ever a local non-promise object representing a remote object.

> I would anticipate that E(issuerP).getAmountOf(paymentP) (and eventually issuerP~.getAmountOf(paymentP)) in one vat would resolve automatically to issuer.getAmountOf(payment) in the issuer vat (and consequently no need to play around with await or E.unwrap), but maybe that there is a reason why it's not possible at all because of language reasons or currently because of temporary limitations

`Unwrap` is necessary because we must say that we want to specifically get the object that the promise resolves to, synchronously, and we can do so safely only because it was a local object. My understanding is that we don't typically want to do this automatically because for security purposes, we want to maintain the "promise barrier" so that our code isn't handing over control to other code. (There is probably a better way to put that.)

**Comment by katelynsills** [comment]
I talked to Mark, and he proposed moving away from `E.unwrap` here: https://github.com/Agoric/agoric-sdk/issues/711

**Comment by katelynsills** [comment]
Closed by #753
